core/src/test/scala/org/broadinstitute/dsde/rawls/billing/BillingProjectOrchestratorSpec.scala

```diff
@@ -291,4 +292,202 @@ class BillingProjectOrchestratorSpec extends AnyFlatSpec {
+                                                   testContext
+    )
+  }
+
+  it should "fail when workspaces attached to the billing project exist" in {
+    val samDAO = mock[SamDAO](RETURNS_SMART_NULLS)
+    val billingProjectName = RawlsBillingProjectName("fake_billing_account_name")
+    when(
+      samDAO.userHasAction(SamResourceTypeNames.billingProject,
+                           billingProjectName.value,
+                           SamBillingProjectActions.deleteBillingProject,
+                           testContext
+      )
+    ).thenReturn(Future.successful(true))
+
+    val billingRepository = mock[BillingRepository]
+    when(billingRepository.failUnlessHasNoWorkspaces(billingProjectName)(executionContext))
+      .thenReturn(
+        Future.failed(
+          new RawlsExceptionWithErrorReport(
+            ErrorReport(
+              StatusCodes.BadRequest,
+              "Project cannot be deleted because it contains workspaces."
+            )
+          )
+        )
+      )
+    // Mock Google project
+    when(billingRepository.getBillingProfileId(billingProjectName)(executionContext))
+      .thenReturn(Future.successful(None))
+
+    val bpo = new BillingProjectOrchestrator(
+      testContext,
+      samDAO,
+      billingRepository,
+      mock[BillingProjectLifecycle](RETURNS_SMART_NULLS),
+      mock[BillingProjectLifecycle](RETURNS_SMART_NULLS),
+      mock[MultiCloudWorkspaceConfig]
+    )
+
+    val ex = intercept[RawlsExceptionWithErrorReport] {
+      Await.result(bpo.deleteBillingProjectV2(billingProjectName), Duration.Inf)
+    }
+
+    assertResult(Some(StatusCodes.BadRequest)) {
+      ex.errorReport.statusCode
+    }
+
+    verify(billingRepository, Mockito.times(1)).failUnlessHasNoWorkspaces(billingProjectName)(executionContext)
+
+  }
+
+  it should "call preDeletionSteps and delete a Google project" in {
+    executeSuccessTest(true)
+  }
+
+  it should "call preDeletionSteps and delete an Azure project" in {
+    executeSuccessTest(false)
+  }
+  def executeSuccessTest(isGoogle: Boolean): Unit = {
+    val samDAO = mock[SamDAO](RETURNS_SMART_NULLS)
+    val billingProjectName = RawlsBillingProjectName("fake_billing_account_name")
+    when(
+      samDAO.userHasAction(SamResourceTypeNames.billingProject,
+                           billingProjectName.value,
+                           SamBillingProjectActions.deleteBillingProject,
+                           testContext
+      )
+    ).thenReturn(Future.successful(true))
+    when(
+      samDAO.deleteResource(SamResourceTypeNames.billingProject, billingProjectName.value, testContext)
+    ).thenReturn(Future.successful())
+
+    val billingRepository = mock[BillingRepository]
+    when(billingRepository.failUnlessHasNoWorkspaces(billingProjectName)(executionContext))
+      .thenReturn(Future.successful())
+    when(billingRepository.deleteBillingProject(billingProjectName)).thenReturn(Future.successful(true))
+
+    // Mock Azure project
```

Comment:
I should remove this comment now that this method can create either type of project.